rmw_cyclonedds_cpp/src/serdata.cpp

```diff
@@ -685,11 +749,351 @@ static std::string get_type_name(const char * type_support_identifier, void * ty
   }
 }
 
+
+extern "C" dds_dynamic_type_descriptor_t get_dynamic_type_descriptor_prim(
+  dds_dynamic_type_kind_t kind, const char *name, uint32_t num_bounds, const uint32_t *bounds, dds_dynamic_type_kind_t type)
+{
+  return{kind,name,{},{},num_bounds,bounds,DDS_DYNAMIC_TYPE_SPEC_PRIM(type),{},};
+}
+
+extern "C" dds_dynamic_type_descriptor_t get_dynamic_type_descriptor(
+  dds_dynamic_type_kind_t kind, const char *name, uint32_t num_bounds, const uint32_t *bounds, dds_dynamic_type_t type)
+{
+  return{kind,name,{},{},num_bounds,bounds,DDS_DYNAMIC_TYPE_SPEC(type),{},};
+}
+
+extern "C" dds_dynamic_member_descriptor_t get_dynamic_member_descriptor_prim(
+  dds_dynamic_type_kind_t type, const char *name)
+{
+  return DDS_DYNAMIC_MEMBER_PRIM(type, name);
+}
+
+extern "C" dds_dynamic_member_descriptor_t get_dynamic_member_descriptor(
+  dds_dynamic_type_t ddt, const char *name)
+{
+  return DDS_DYNAMIC_MEMBER(ddt, name);
+}
+
+
+template<typename MemberType>
+static void dynamic_type_add_array(
+  dds_dynamic_type_t * dstruct, dds_entity_t dds_ppant, const MemberType * member, const dds_dynamic_type_kind_t type)
+{
+  dds_dynamic_type_t ddt;
+
+  if (member->array_size_) 
+  { 
+    if (!member->is_upper_bound_)
+    {
+      ddt = dds_dynamic_type_create(dds_ppant, 
+        get_dynamic_type_descriptor_prim(DDS_DYNAMIC_ARRAY, member->name_, 1, (uint32_t*)(&member->array_size_), type));
+    } else {
+      ddt = dds_dynamic_type_create(dds_ppant, 
+        get_dynamic_type_descriptor_prim(DDS_DYNAMIC_SEQUENCE, member->name_, 1, (uint32_t*)(&member->array_size_), type)); 
+    }
+  } else {
+    ddt = dds_dynamic_type_create(dds_ppant,
+       get_dynamic_type_descriptor_prim(DDS_DYNAMIC_SEQUENCE, member->name_, 0, nullptr, type)); 
+  }
+
+  dds_return_t ret = dds_dynamic_type_add_member(dstruct, get_dynamic_member_descriptor(ddt, member->name_));
+  if (ret != DDS_RETCODE_OK)
+      RCUTILS_LOG_ERROR("failed to add type member sequence/array with prim type");
+}
+
+template<typename MemberType>
```

Comment:
I must be overlooking something obvious, because it looks like this is so close to the function above that it might as well not exist!

(But then surely the compiler would reject it. So I must be the one who is wrong here.)

Reply:
We still need two instances of the function. one for adding primitive types, one for more complex types.